plugins/aver-hid/fu-aver-safeisp-device.c

```diff
@@ -0,0 +1,448 @@
+	    req,
+	    FU_AVER_SAFEISP_CUSTOM_CMD_UPLOAD_COMPARE_CHECKSUM);
+	fu_struct_aver_safeisp_req_set_custom_parm0(req, param0);
+	fu_struct_aver_safeisp_req_set_custom_parm1(req, param1);
+	if (!fu_aver_safeisp_device_transfer(self, req, res, error))
+		return FALSE;
+	if (!fu_struct_aver_safeisp_res_validate(res->data, res->len, 0x0, error))
+		return FALSE;
+	if (fu_struct_aver_safeisp_req_get_custom_cmd(res) != FU_AVER_SAFEISP_ACK_STATUS_SUCCESS)
+		return FALSE;
+	return TRUE;
+}
+
+static gboolean
+fu_aver_safeisp_device_update(FuAverSafeispDevice *self, gsize param0, gsize param1, GError **error)
+{
+	g_autoptr(GByteArray) req = fu_struct_aver_safeisp_req_new();
+
+	fu_struct_aver_safeisp_req_set_custom_cmd(req, FU_AVER_SAFEISP_CUSTOM_CMD_UPDATE_START);
+	fu_struct_aver_safeisp_req_set_custom_parm0(req, param0);
+	fu_struct_aver_safeisp_req_set_custom_parm1(req, param1);
+	if (!fu_aver_safeisp_device_transfer(self, req, NULL, error))
+		return FALSE;
+	return TRUE;
+}
+
+static gboolean
+fu_aver_safeisp_device_write_firmware(FuDevice *device,
+				      FuFirmware *firmware,
+				      FuProgress *progress,
+				      FwupdInstallFlags flags,
+				      GError **error)
+{
+	FuAverSafeispDevice *self = FU_AVER_SAFEISP_DEVICE(device);
+	g_autoptr(FuArchive) archive = NULL;
+	g_autoptr(GBytes) fw = NULL;
+	g_autoptr(GBytes) cx3_fw = NULL;
+	g_autoptr(GBytes) m12_fw = NULL;
+	gsize cx3_fw_size;
+	gsize m12_fw_size;
+	const guint8 *cx3_fw_buf;
+	const guint8 *m12_fw_buf;
+	guint32 cx3_checksum = 0;
+	guint32 m12_checksum = 0;
+	g_autoptr(FuChunkArray) chunks = NULL;
+
+	/* progress */
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_BUSY, 1, NULL);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_WRITE, 5, NULL);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_VERIFY, 1, NULL);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_BUSY, 1, NULL);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_WRITE, 58, NULL);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_VERIFY, 34, NULL);
+
+	/* get default image */
+	fw = fu_firmware_get_bytes(firmware, error);
+	if (fw == NULL)
+		return FALSE;
+
+	/* decompress */
+	archive = fu_archive_new(fw, FU_ARCHIVE_FLAG_NONE, error);
+	if (archive == NULL)
+		return FALSE;
+	cx3_fw = fu_archive_lookup_by_fn(archive, "update/cx3uvc.img", error);
+	if (cx3_fw == NULL)
+		return FALSE;
+	m12_fw = fu_archive_lookup_by_fn(archive, "update/RS_M12MO.bin", error);
+	if (m12_fw == NULL)
+		return FALSE;
+
+	/* CX3 fw file size should be less than 256KB */
+	cx3_fw_buf = g_bytes_get_data(cx3_fw, &cx3_fw_size);
+	if (cx3_fw_size > 256 * 1024)
+		return FALSE;
+	/* calculate CX3 firmware checksum */
+	for (guint i = 0; i < cx3_fw_size; i++)
+		cx3_checksum += cx3_fw_buf[i];
+	/* M12 fw file size should be less than 3MB */
+	m12_fw_buf = g_bytes_get_data(m12_fw, &m12_fw_size);
+	if (m12_fw_size > 3 * 1024 * 1024)
+		return FALSE;
```

Comment:
same, needs GError set

Reply:
same here with including the size - "in the field" we need these failures to be debuggable.